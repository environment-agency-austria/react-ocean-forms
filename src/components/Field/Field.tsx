/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as React from 'react';

import { getDeepValue } from '../../utils';
import { IFormContext, TFieldValues } from '../FormContext';
import { IValidationArgs, IValidationState, withValidation } from '../withValidation';
import { IFieldComponentFieldProps, IFieldComponentMeta, IFieldProps, IValueMeta, TFieldValue } from './Field.types';

interface IContextMeta extends IValueMeta {
  defaultValue?: TFieldValue;
  externalValue?: TFieldValue;
}

interface IFieldState {
  touched: boolean;
  dirty: boolean;
  value: TFieldValue;
  contextMeta: IContextMeta;
}

interface IMetaChanges {
  defaultValue: boolean;
  externalValue: boolean;
  meta: boolean;
}

/**
 * Wrapper for input fields managed by
 * the form component
 */
export class BaseField extends React.Component<IFieldProps, IFieldState> {
  public static displayName: string = 'Field';

  // tslint:disable-next-line:typedef
  public static defaultProps = {
    onChange: (): void => undefined,
    onBlur: (): void => undefined,
    getDisplayValue: (value: TFieldValue): TFieldValue => value,
    getSubmitValue: (value: TFieldValue): TFieldValue => value,
  };

  constructor(props: IFieldProps) {
    super(props);

    const {
      fullName,
      label,
      context,
      validation: {
        update: updateValidation,
      },
    } = props;

    this.checkFormContext();

    context.registerField(
      fullName,
      {
        label,

        updateValidation,
        validate: this.validate,
        reset: this.reset,
        getValue: this.getValue,
      },
    );

    this.state = {
      touched: false,
      dirty: false,
      value: BaseField.callGetDisplayValue(props, ''),
      contextMeta: {
        disabled: false,
        plaintext: false,
      },
    };
  }

  /**
   * Changes the field value based on changes in Form.defaultValues,
   * Form.values, Field.defaultValue or Field.value. The Field props
   * always override the Form props. Changes in defaultValue will only
   * update the field value, if there is no values or value prop present
   * and if the field hasn't been touched.
   */
  public static getDerivedStateFromProps(nextProps: IFieldProps, prevState: IFieldState): IFieldState | null {
    const {
      context: {
        disabled: newDisabled,
        plaintext: newPlaintext,
      },
    } = nextProps;

    // Get the default value from Field.defaultValue or Form.defaultValues
    const defaultValue = BaseField.getDefaultValue(nextProps);
    // Get the "external" value from Field.value or Form.values
    const externalValue = BaseField.getExternalValue(nextProps);

    // Get changes in default or external value and meta information (plaintext, disabled)
    // Meta information must trigger a state change as well, because they are passed to
    // Field.getDisplayValue and could result in a different return value of that function
    const changes = BaseField.getPropChanges(nextProps, prevState, defaultValue, externalValue);
    // Get the new value based on the information above, or undefined if we don't need any
    // value changes.
    const propValue = BaseField.getPropValue(defaultValue, externalValue, changes);

    // Update the Field state if needed, also remember the current prop values so we can
    // detect changes in the next call of getDerivedStateFromProps.
    if (propValue !== undefined) {
      return ({
        value: BaseField.callGetDisplayValue(nextProps, propValue),
        touched: false,
        dirty: false,
        contextMeta: {
          defaultValue,
          externalValue,
          disabled: newDisabled,
          plaintext: newPlaintext,
        },
      });
    }

    return null;
  }

  /**
   * Correctly calls the getDisplayValue callback
   * @param props Field props
   * @param value Field value
   */
  private static callGetDisplayValue(props: IFieldProps, value: TFieldValue | undefined): TFieldValue {
    const { getDisplayValue, context } = props;

    return getDisplayValue(
      value || '',
      BaseField.getValueMeta(context),
    );
  }

  /**
   * Calculates the new state value based on the parameters - it will return
   * the externalValue if it exists and has changes, or the default value
   * otherwise. If no state update is needed it will return undefined. Meta
   * changes will trigger a state change as well.
   * @param defaultValue Default value
   * @param externalValue External value
   * @param changes Changes object
   */
  private static getPropValue(
    defaultValue: TFieldValue | undefined, externalValue: TFieldValue | undefined, changes: IMetaChanges,
  ): TFieldValue | undefined {
    const hasExternalValue = externalValue !== undefined;

    if (hasExternalValue && (changes.externalValue || changes.meta)) {
      return externalValue;
    }

    if (!hasExternalValue && (changes.defaultValue || changes.meta)) {
      return defaultValue;
    }

    return undefined;
  }

  /**
   * Detects changes in the defaultValue, the externalValue and in
   * context.disabled or context.plaintext (merged as meta prop)
   * @param props Field props
   * @param state Field state
   * @param defaultValue Default value
   * @param externalValue External value
   */
  private static getPropChanges(
    props: IFieldProps, state: IFieldState, defaultValue: TFieldValue | undefined, externalValue: TFieldValue | undefined,
  ): IMetaChanges {
    const {
      context: {
        disabled: newDisabled,
        plaintext: newPlaintext,
      },
    } = props;

    const {
      contextMeta: {
        defaultValue: oldDefaultValue,
        externalValue: oldExternalValue,
        disabled: oldDisabled,
        plaintext: oldPlaintext,
      },
      touched,
    } = state;

    return {
      defaultValue: !touched && defaultValue !== oldDefaultValue,
      externalValue: externalValue !== oldExternalValue,
      meta: newDisabled !== oldDisabled || newPlaintext !== oldPlaintext,
    };
  }

  /**
   * Returns the local value if existing, otherwise tries to
   * extract the correct context value based on the fullName.
   * @param localValue Local value
   * @param contextValue Context value
   * @param fullName Field.fullName
   */
  private static getLocalOverridenValue(
    localValue: TFieldValue | undefined, contextValue: TFieldValues | undefined, fullName: string,
  ): TFieldValue | undefined {
    return localValue || getDeepValue(fullName, contextValue);
  }

  /**
   * Returns either the Field.value or the correct value from
   * Form.values
   * @param props Field props
   */
  private static getExternalValue(props: IFieldProps): TFieldValue | undefined {
    const { fullName, context: { values }, value } = props;

    return BaseField.getLocalOverridenValue(value, values, fullName);
  }

  /**
   * Returns either the Field.defaultValue or the correct default value
   * from Form.defaultValues
   * @param props Field props
   */
  private static getDefaultValue(props: IFieldProps): TFieldValue | undefined {
    const { fullName, context: { defaultValues }, defaultValue } = props;

    return BaseField.getLocalOverridenValue(defaultValue, defaultValues, fullName);
  }

  /**
   * Returns a meta object for the value lifecycle hooks
   * @param context Form context
   */
  private static getValueMeta(context: IFormContext): IValueMeta {
    return {
      disabled: context.disabled,
      plaintext: context.plaintext,
    };
  }

  /**
   * Unregisters the field from the form
   */
  public componentWillUnmount(): void {
    const { context, fullName } = this.props;
    context.unregisterField(fullName);
  }

  /**
   * Returns the current field value
   */
  private getValue = (): TFieldValue => {
    const { value } = this.state;
    const { getSubmitValue, context } = this.props;

    return getSubmitValue(
      value,
      BaseField.getValueMeta(context),
    );
  }

  /**
   * Returns the correct asyncValidateOnChange setting,
   * where the field setting takes priorty over the
   * form setting
   */
  private getAsyncValidateOnChangeSetting(): boolean {
    const {
      asyncValidateOnChange: propChange,
      context: { asyncValidateOnChange: contextChange },
    } = this.props;

    return propChange === undefined ? contextChange : propChange;
  }

  /**
   * Checks if the Field is inside a valid form context
   * and throws an user friendly error if not
   */
  private checkFormContext(): void {
    const { context, fullName } = this.props;
    if (!context || typeof context.registerField !== 'function') {
      throw new Error(
        `Could not find a form context for field "${fullName}". `
        + 'Fields can only be used inside a Form tag.',
      );
    }
  }

  /**
   * Resets the field to its default state
   */
  private reset = (): void => {
    const { contextMeta: { externalValue, defaultValue } } = this.state;
    const { validation, onChange } = this.props;

    const value = BaseField.callGetDisplayValue(
      this.props,
      externalValue || defaultValue,
    );

    this.setState({
      value,
      touched: false,
      dirty: false,
    });

    validation.reset();
    onChange(value);
  }

  /**
   * Validates the field
   * @param args Validation arguments
   */
  private validate = async (args?: Partial<IValidationArgs>): Promise<IValidationState> => {
    const { value } = this.state;
    const { validation: { validate }, getSubmitValue, context } = this.props;

    return validate(
      getSubmitValue(
        value,
        BaseField.getValueMeta(context),
      ),
      args,
    );
  }

  /**
   * Handles the change event of an input field -
   * triggers any validation if needed and updates
   * the field state accordingly.
   * @param event Event object
   */
  private handleFieldChanged = (event: React.ChangeEvent<HTMLInputElement>): void => {
    const { value } = event.target;
    const {
      fullName,
      context,
      validation: { validate },
      getSubmitValue,
      onChange,
    } = this.props;

    this.setState({
      value,
      dirty: true,
      touched: true,
    });

    const asyncValidateOnChange = this.getAsyncValidateOnChangeSetting();
    const submitValue = getSubmitValue(
      value,
      BaseField.getValueMeta(context),
    );

    validate(
      submitValue,
      { checkAsync: asyncValidateOnChange },
    );

    context.notifyFieldEvent(fullName, 'change', submitValue);
    onChange(submitValue);
  }

  /**
   * Handles the blur event of an input field -
   * triggers any validation if needed and updates
   * the field state accordingly.
   */
  private handleFieldBlurred = (): void => {
    const {
      fullName,
      validation: { validate },
      context,
      onBlur,
      getSubmitValue,
    } = this.props;
    const { value, dirty } = this.state;

    const asyncValidateOnChange = this.getAsyncValidateOnChangeSetting();
    const submitValue = getSubmitValue(
      value,
      BaseField.getValueMeta(context),
    );

    if (dirty && !asyncValidateOnChange) { validate(submitValue); }
    context.notifyFieldEvent(fullName, 'blur');
    onBlur();
  }

  /**
   * Creates the properties that can be directly
   * mapped to the input component (e.g. html input)
   */
  private createFieldProps(): IFieldComponentFieldProps {
    const { context: { disabled }, fullName } = this.props;
    const { value } = this.state;

    return {
      value,
      disabled,
      id: fullName,
      name: fullName,
      onChange: this.handleFieldChanged,
      onBlur: this.handleFieldBlurred,
    };
  }

  /**
   * Creates the meta properties with meta information
   * used by the input component
   */
  private createMetaProps(): IFieldComponentMeta {
    const {
      context: {
        stringFormatter,
        plaintext,
      },
      validation: {
        valid,
        error,
        isValidating,
      },
    } = this.props;

    const { touched } = this.state;

    return {
      valid,
      error,
      isValidating,
      touched,
      stringFormatter,
      plaintext,
    };
  }

  // tslint:disable-next-line:member-ordering
  public render(): JSX.Element {
    const {
      // tslint:disable-next-line:naming-convention
      component: InputComponent,
      context,
      fullName,
      validation,
      onChange,
      onBlur,
      defaultValue,
      value,
      name,
      asyncValidateOnChange,
      getDisplayValue,
      getSubmitValue,
      ...attributes
    } = this.props;

    const field = this.createFieldProps();
    const meta = this.createMetaProps();

    return (
      <InputComponent
        {...attributes}
        field={field}
        meta={meta}
      />
    );
  }
}

export const Field = withValidation(BaseField);
