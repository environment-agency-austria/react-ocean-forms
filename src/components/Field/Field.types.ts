/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { TSTringFormatter } from '../../utils';
import { IValidationState } from '../../hooks';
import { IValidationProps } from '../withValidation';

/**
 * Type that defines which values a field could hold
 */
export type TBasicFieldValue = string | boolean | number | object;
/**
 * Type definition for getDisplayValue and getSubmitValue callbacks
 */
export type TValueCallback = ((value: TBasicFieldValue, meta: IValueMeta) => TBasicFieldValue);

/**
 * Information about the current meta state
 * of the Field
 */
export interface IValueMeta {
  /**
   * True, if the field is disabled
   */
  disabled: boolean;
  /**
   * True, if the field is in plaintext mode
   */
  plaintext: boolean;
}

/**
 * Basic props for the field component
 */
export interface IBaseFieldProps {
  /**
   * Field name
   */
  name: string;
  /**
   * Label (string or message id)
   */
  label: string;
  /**
   * Optional default value
   */
  defaultValue?: TBasicFieldValue;
  /**
   * Optional value
   */
  value?: TBasicFieldValue;
  /**
   * True, if the async validators should be triggered
   * on a change event
   */
  asyncValidateOnChange?: boolean;
  /**
   * Called, when the field is loading its value from the forms
   * default values. Must return the value to display.
   */
  getDisplayValue?: TValueCallback;
  /**
   * Called, when the field is submitting its value to the form.
   * Must return the value to submit.
   */
  getSubmitValue?: TValueCallback;
  /**
   * Disables this field.
   */
  disabled?: boolean;
  /**
   * Puts the field in plaintext mode.
   */
  plaintext?: boolean;
  /**
   * Triggered on field blur.
   */
  onBlur?(): void;
  /**
   * Triggered on field value change.
   * @param value Current field value
   */
  onChange?(value: TBasicFieldValue): void;
}

/**
 * Props for the Field component
 */
export interface IFieldProps extends IBaseFieldProps, IValidationProps {
  /**
   * Render prop for the input element
   * @param field Props designed to be passed to the field as is @see IFieldComponentFieldProps
   * @param meta Meta information about the field state
   */
  render(field: IFieldComponentFieldProps, meta: IFieldComponentMeta): JSX.Element;
}

/**
 * Event for field value changes
 */
export interface IFieldChangedEvent {
  target: {
    value: TBasicFieldValue;
  };
}

/**
 * Props for the actual html input of a Field
 * input component. Designed to be passed to the
 * html input as-is.
 */
export interface IFieldComponentFieldProps {
  /**
   * Html id
   */
  id: string;
  /**
   * Field name
   */
  name: string;
  /**
   * Field value
   */
  value: TBasicFieldValue;
  /**
   * Disabled state
   */
  disabled: boolean;
  /**
   * OnChange handler
   * @param event Change event
   */
  onChange(event: IFieldChangedEvent): void;
  /**
   * OnBlur handler
   */
  onBlur(): void;
}

/**
 * Meta informations about the current field state
 */
export interface IFieldComponentMeta extends IValidationState {
  /**
   * True, if the user has changed the value
   */
  touched: boolean;
  /**
   * String formatter method
   */
  stringFormatter: TSTringFormatter;
  /**
   * True if the field is in plaintext mode
   */
  plaintext: boolean;
}

/**
 * Props for the input component of a Field
 */
export interface IFieldComponentProps {
  /**
   * Props for the actual html input, designed
   * to be passed as-is
   */
  field: IFieldComponentFieldProps;
  /**
   * Meta informations about the field state
   */
  meta: IFieldComponentMeta;
  /**
   * Label (string or message id)
   */
  label: string;
}
