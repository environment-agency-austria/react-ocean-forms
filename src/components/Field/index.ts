export * from './Field';
export * from './Field.types';
