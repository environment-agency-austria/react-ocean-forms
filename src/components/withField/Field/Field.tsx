/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';

import { getDeepValue } from '../../../utils';
import { IFieldValues, IFormContext } from '../../FormContext';
import { IValidationArgs, IValidationState, withValidation } from '../../withValidation';
import {
  IFieldChangedEvent,
  IFieldComponentFieldProps,
  IFieldComponentMeta,
  IFieldProps,
  IValueMeta,
  TBasicFieldValue,
} from './Field.types';

interface IContextMeta extends IValueMeta {
  defaultValue?: TBasicFieldValue;
  externalValue?: TBasicFieldValue;
}

interface IFieldState {
  touched: boolean;
  dirty: boolean;
  value: TBasicFieldValue;
  contextMeta: IContextMeta;
}

interface IMetaChanges {
  defaultValue: boolean;
  externalValue: boolean;
  meta: boolean;
}

/**
 * Wrapper for input fields managed by
 * the form component
 */
export class BaseField extends React.Component<IFieldProps, IFieldState> {
  public static displayName: string = 'Field';

  constructor(props: IFieldProps) {
    super(props);

    const {
      fullName,
      label,
      context,
      validation: {
        update: updateValidation,
      },
    } = props;

    this.checkFormContext();

    context.registerField(
      fullName,
      {
        label,

        updateValidation,
        validate: this.validate,
        reset: this.reset,
        getValue: this.getValue,
      },
    );

    this.state = {
      touched: false,
      dirty: false,
      value: BaseField.callGetDisplayValue(props, ''),
      contextMeta: {
        disabled: false,
        plaintext: false,
      },
    };
  }

  /**
   * Changes the field value based on changes in Form.defaultValues,
   * Form.values, Field.defaultValue or Field.value. The Field props
   * always override the Form props. Changes in defaultValue will only
   * update the field value, if there is no values or value prop present
   * and if the field hasn't been touched.
   */
  public static getDerivedStateFromProps(nextProps: IFieldProps, prevState: IFieldState): IFieldState | null {
    const {
      context: {
        disabled: newDisabled,
        plaintext: newPlaintext,
      },
    } = nextProps;

    // Get the default value from Field.defaultValue or Form.defaultValues
    const defaultValue = BaseField.getDefaultValue(nextProps);
    // Get the "external" value from Field.value or Form.values
    const externalValue = BaseField.getExternalValue(nextProps);

    // Get changes in default or external value and meta information (plaintext, disabled)
    // Meta information must trigger a state change as well, because they are passed to
    // Field.getDisplayValue and could result in a different return value of that function
    const changes = BaseField.getPropChanges(nextProps, prevState, defaultValue, externalValue);
    // Get the new value based on the information above, or undefined if we don't need any
    // value changes.
    const propValue = BaseField.getPropValue(defaultValue, externalValue, changes);

    // Update the Field state if needed, also remember the current prop values so we can
    // detect changes in the next call of getDerivedStateFromProps.
    if (propValue !== undefined) {
      return ({
        value: BaseField.callGetDisplayValue(nextProps, propValue),
        touched: false,
        dirty: false,
        contextMeta: {
          defaultValue,
          externalValue,
          disabled: newDisabled,
          plaintext: newPlaintext,
        },
      });
    }

    return null;
  }

  /**
   * Correctly calls the getDisplayValue callback
   * @param props Field props
   * @param value Field value
   */
  private static callGetDisplayValue(props: IFieldProps, value: TBasicFieldValue | undefined): TBasicFieldValue {
    const {
      context,
      getDisplayValue = (val: TBasicFieldValue): TBasicFieldValue => val,
    } = props;

    return getDisplayValue(
      value === undefined ? '' : value,
      BaseField.getValueMeta(context),
    );
  }

  /**
   * Calculates the new state value based on the parameters - it will return
   * the externalValue if it exists and has changes, or the default value
   * otherwise. If no state update is needed it will return undefined. Meta
   * changes will trigger a state change as well.
   * @param defaultValue Default value
   * @param externalValue External value
   * @param changes Changes object
   */
  private static getPropValue(
    defaultValue: TBasicFieldValue | undefined, externalValue: TBasicFieldValue | undefined, changes: IMetaChanges,
  ): TBasicFieldValue | undefined {
    const hasExternalValue = externalValue !== undefined;

    if (hasExternalValue && (changes.externalValue || changes.meta)) {
      return externalValue;
    }

    if (!hasExternalValue && (changes.defaultValue || changes.meta)) {
      return defaultValue;
    }

    return undefined;
  }

  /**
   * Detects changes in the defaultValue, the externalValue and in
   * context.disabled or context.plaintext (merged as meta prop)
   * @param props Field props
   * @param state Field state
   * @param defaultValue Default value
   * @param externalValue External value
   */
  private static getPropChanges(
    props: IFieldProps, state: IFieldState, defaultValue: TBasicFieldValue | undefined, externalValue: TBasicFieldValue | undefined,
  ): IMetaChanges {
    const {
      context: {
        disabled: newDisabled,
        plaintext: newPlaintext,
      },
    } = props;

    const {
      contextMeta: {
        defaultValue: oldDefaultValue,
        externalValue: oldExternalValue,
        disabled: oldDisabled,
        plaintext: oldPlaintext,
      },
      touched,
    } = state;

    return {
      defaultValue: !touched && defaultValue !== oldDefaultValue,
      externalValue: externalValue !== oldExternalValue,
      meta: newDisabled !== oldDisabled || newPlaintext !== oldPlaintext,
    };
  }

  /**
   * Returns the local value if existing, otherwise tries to
   * extract the correct context value based on the fullName.
   * @param localValue Local value
   * @param contextValue Context value
   * @param fullName Field.fullName
   */
  private static getLocalOverridenValue(
    localValue: TBasicFieldValue | undefined, contextValue: Partial<IFieldValues> | undefined, fullName: string,
  ): TBasicFieldValue | undefined {
    return localValue === undefined ? getDeepValue(fullName, contextValue) : localValue;
  }

  /**
   * Returns either the Field.value or the correct value from
   * Form.values
   * @param props Field props
   */
  private static getExternalValue(props: IFieldProps): TBasicFieldValue | undefined {
    const { fullName, context: { values }, value } = props;

    return BaseField.getLocalOverridenValue(value, values, fullName);
  }

  /**
   * Returns either the Field.defaultValue or the correct default value
   * from Form.defaultValues
   * @param props Field props
   */
  private static getDefaultValue(props: IFieldProps): TBasicFieldValue | undefined {
    const { fullName, context: { defaultValues }, defaultValue } = props;

    return BaseField.getLocalOverridenValue(defaultValue, defaultValues, fullName);
  }

  /**
   * Returns a meta object for the value lifecycle hooks
   * @param context Form context
   */
  private static getValueMeta(context: IFormContext): IValueMeta {
    return {
      disabled: context.disabled,
      plaintext: context.plaintext,
    };
  }

  /**
   * Unregisters the field from the form
   */
  public componentWillUnmount(): void {
    const { context, fullName } = this.props;
    context.unregisterField(fullName);
  }

  /**
   * Returns the current field value
   */
  private getValue = (): TBasicFieldValue => {
    const { value } = this.state;

    return this.getSubmitValue(value);
  }

  private getSubmitValue = (value: TBasicFieldValue): TBasicFieldValue => {
    const {
      context,
      getSubmitValue = (val: TBasicFieldValue): TBasicFieldValue => val,
    } = this.props;

    return getSubmitValue(
      value,
      BaseField.getValueMeta(context),
    );
  }

  /**
   * Returns the correct asyncValidateOnChange setting,
   * where the field setting takes priorty over the
   * form setting
   */
  private getAsyncValidateOnChangeSetting(): boolean {
    const {
      asyncValidateOnChange: propChange,
      context: { asyncValidateOnChange: contextChange },
    } = this.props;

    return propChange === undefined ? contextChange : propChange;
  }

  /**
   * Checks if the Field is inside a valid form context
   * and throws an user friendly error if not
   */
  private checkFormContext(): void {
    const { context, fullName } = this.props;
    if (context === undefined || typeof context.registerField !== 'function') {
      throw new Error(
        `Could not find a form context for field "${fullName}". `
        + 'Fields can only be used inside a Form tag.',
      );
    }
  }

  /**
   * Resets the field to its default state
   */
  private reset = (): void => {
    const { contextMeta: { externalValue, defaultValue } } = this.state;
    const { validation, onChange } = this.props;

    const value = BaseField.callGetDisplayValue(
      this.props,
      externalValue === undefined ? defaultValue : externalValue,
    );

    this.setState({
      value,
      touched: false,
      dirty: false,
    });

    validation.reset();
    if (onChange !== undefined) { onChange(value); }
  }

  /**
   * Validates the field
   * @param args Validation arguments
   */
  private validate = async (args?: Partial<IValidationArgs>): Promise<IValidationState> => {
    const { value } = this.state;
    const { validation: { validate } } = this.props;

    return validate(
      this.getSubmitValue(value),
      args,
    );
  }

  /**
   * Handles the change event of an input field -
   * triggers any validation if needed and updates
   * the field state accordingly.
   * @param event Event object
   */
  private handleFieldChanged = (event: IFieldChangedEvent): void => {
    const { value } = event.target;
    const {
      fullName,
      context,
      validation: { validate },
      onChange,
    } = this.props;

    this.setState({
      value,
      dirty: true,
      touched: true,
    });

    const asyncValidateOnChange = this.getAsyncValidateOnChangeSetting();
    const submitValue = this.getSubmitValue(value);

    void validate(
      submitValue,
      { checkAsync: asyncValidateOnChange },
    );

    context.notifyFieldEvent(fullName, 'change', submitValue);
    if (onChange !== undefined) { onChange(submitValue); }
  }

  /**
   * Handles the blur event of an input field -
   * triggers any validation if needed and updates
   * the field state accordingly.
   */
  private handleFieldBlurred = (): void => {
    const {
      fullName,
      validation: { validate },
      context,
      onBlur,
    } = this.props;
    const { value, dirty } = this.state;

    const asyncValidateOnChange = this.getAsyncValidateOnChangeSetting();
    const submitValue = this.getSubmitValue(value);

    if (dirty && !asyncValidateOnChange) { void validate(submitValue); }
    context.notifyFieldEvent(fullName, 'blur');
    if (onBlur !== undefined) { onBlur(); }
  }

  /**
   * Creates the properties that can be directly
   * mapped to the input component (e.g. html input)
   */
  private createFieldProps(): IFieldComponentFieldProps {
    const { context: { disabled }, fullName } = this.props;
    const { value } = this.state;

    return {
      value,
      disabled,
      id: fullName,
      name: fullName,
      onChange: this.handleFieldChanged,
      onBlur: this.handleFieldBlurred,
    };
  }

  /**
   * Creates the meta properties with meta information
   * used by the input component
   */
  private createMetaProps(): IFieldComponentMeta {
    const {
      context: {
        stringFormatter,
        plaintext,
      },
      validation: {
        valid,
        error,
        isValidating,
        isRequired,
      },
    } = this.props;

    const { touched } = this.state;

    return {
      valid,
      error,
      isValidating,
      isRequired,
      touched,
      stringFormatter,
      plaintext,
    };
  }

  // tslint:disable-next-line:member-ordering
  public render(): JSX.Element {
    const { render } = this.props;

    return render(
      this.createFieldProps(),
      this.createMetaProps(),
    );
  }
}

export const Field = withValidation(BaseField);
