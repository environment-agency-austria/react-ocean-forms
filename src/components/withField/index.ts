export * from './withField';
export * from './Field/Field.types';
