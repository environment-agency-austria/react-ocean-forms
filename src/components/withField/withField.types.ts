/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { IFieldComponentFieldProps, IFieldComponentMeta } from '../../hooks';

/**
 * Props for the input component of a Field
 */
export interface IFieldComponentProps {
  /**
   * Props for the actual html input, designed
   * to be passed as-is
   */
  field: IFieldComponentFieldProps;
  /**
   * Meta informations about the field state
   */
  meta: IFieldComponentMeta;
  /**
   * Label (string or message id)
   */
  label: string;
}
