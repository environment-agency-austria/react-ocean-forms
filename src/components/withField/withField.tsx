/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';

import { getDisplayName, PropsOf, Subtract } from '../../utils';
import { useField, IBaseFieldProps, IUseValidationArgs } from '../../hooks';
import { IFieldComponentProps } from './withField.types';

type WrappedValidatedComponentProps<TComp> =
  Subtract<JSX.LibraryManagedAttributes<TComp, PropsOf<TComp>>, IFieldComponentProps> & IBaseFieldProps & IUseValidationArgs;

/**
 * Higher order component for validation
 * @deprecated Deprecated in favour of `useField` hook
 */
export const withField = <TComp extends React.ComponentType<TProps>, TProps extends IFieldComponentProps = PropsOf<TComp>>
(component: TComp): React.ComponentType<WrappedValidatedComponentProps<TComp>> => {

  const CastedComponent = component as React.ComponentType<TProps>;

  type IWrappedProps = WrappedValidatedComponentProps<TComp>;

  const ValidatedComponent: React.SFC<IWrappedProps> = (props: IWrappedProps): JSX.Element => {
    const { fieldProps, metaProps } = useField(props);

    return (
      // @ts-ignore
      <CastedComponent field={fieldProps} meta={metaProps} {...props} />
    );
  };
  ValidatedComponent.displayName = `withField(${getDisplayName(component)})`;

  return ValidatedComponent;
};
