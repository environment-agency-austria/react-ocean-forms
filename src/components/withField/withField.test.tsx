import React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';

import { IFieldComponentFieldProps, IFieldComponentMeta, IFieldComponentProps } from './Field/Field.types';
import { withField } from './withField';

describe('withField', () => {
  interface ISetupArgs {
    props?: Partial<IFieldComponentProps>;
  }

  interface ISetupResult {
    wrapper: ShallowWrapper;
    renderProp(field: IFieldComponentFieldProps, meta: IFieldComponentMeta): JSX.Element;
  }

  const setup = ({
    props,
  }: Partial<ISetupArgs> = {}): ISetupResult => {
    // tslint:disable-next-line:naming-convention
    const TestComponent = (): JSX.Element => (<div id="test-component" />);
    // tslint:disable-next-line:naming-convention
    const WrappedComponent = withField(TestComponent);

    const wrapper = shallow((
      <WrappedComponent
        name="mock-item"
        label="Mock Item"
        {...props}
      />
    ));

    const renderProp = wrapper.prop('render') as ((field: IFieldComponentFieldProps, meta: IFieldComponentMeta) => JSX.Element);

    return {
      wrapper,
      renderProp,
    };
  };

  it('should render without error', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  describe('render prop', () => {
    it('should render without error', () => {
      const { renderProp } = setup();
      const wrapper = shallow(
        renderProp(
          {
            disabled: false,
            id: 'mock-item',
            name: 'mock-item',
            value: '',
            onChange: jest.fn(),
            onBlur: jest.fn(),
          },
          {
            error: null,
            isValidating: false,
            isRequired: false,
            plaintext: false,
            stringFormatter: jest.fn(),
            touched: false,
            valid: true,
          },
        ),
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
});
