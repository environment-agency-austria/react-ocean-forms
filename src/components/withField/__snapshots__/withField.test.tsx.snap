// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withField render prop should render without error 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TestComponent
    field={
      Object {
        "disabled": false,
        "id": "mock-item",
        "name": "mock-item",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "value": "",
      }
    }
    label="Mock Item"
    meta={
      Object {
        "error": null,
        "isValidating": false,
        "plaintext": false,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      }
    }
    name="mock-item"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "id": "test-component",
    },
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "id": "test-component",
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`withField should render without error 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withField(TestComponent)
    label="Mock Item"
    name="mock-item"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "label": "Mock Item",
      "name": "mock-item",
      "render": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "label": "Mock Item",
        "name": "mock-item",
        "render": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
